### ISSUE TITLE
Error to train on my data 

### ISSUE THREAD

**Issue by bleedingfight** [tensorflow/models#4524]
My environment;
cuda V9.0.176,cudnn 7.0,tensorflow-gpu 1.8,Titan X x8
My data:20000 num picture,2labels,object is lane(4 pixle width ),other is background.Every picture size is[512,512],segmentation annotation picture is grayscale picture,lane pix is 1,other's(background) is 0.
my alter include:

    splits_to_sizes={
        'train': 18000,
        'trainval': 20000,
        'val': 2000,
    },
    num_classes=2,
    ignore_label=255,
)```
and add:
```_DATASETS_INFORMATION = {
    'cityscapes': _CITYSCAPES_INFORMATION,
    'pascal_voc_seg': _PASCAL_VOC_SEG_INFORMATION,
    'lane_seg': _LANE_SEG_INFORMATION,
    'ade20k': _ADE20K_INFORMATION,
}

train:

NUM_ITERATIONS=10000
python "${WORK_DIR}"/train.py \
  --logtostderr \
  --initialize_last_layer=False \
  --num_clones=1 \
  --last_layers_contain_logits_only=False \
  --dataset='lane_seg' \
  --train_split="trainval" \
  --model_variant="xception_65" \
  --atrous_rates=6 \
  --atrous_rates=12 \
  --atrous_rates=18 \
  --output_stride=16 \
  --decoder_output_stride=4 \
  --train_crop_size=513 \
  --train_crop_size=513 \
  --train_batch_size=4 \
  --training_number_of_steps="${NUM_ITERATIONS}" \
  --fine_tune_batch_norm=true \
  --tf_initial_checkpoint="${INIT_FOLDER}/deeplabv3_pascal_train_aug/model.ckpt" \
  --train_logdir="${TRAIN_LOGDIR}" \
  --dataset_dir="${LANE_DATASET}"

python "${WORK_DIR}"/eval.py \
  --logtostderr \
  --eval_split="val" \
  --dataset="lane_seg" \
  --model_variant="xception_65" \
  --atrous_rates=6 \
  --atrous_rates=12 \
  --atrous_rates=18 \
  --output_stride=16 \
  --decoder_output_stride=4 \
  --eval_crop_size=513 \
  --eval_crop_size=513 \
  --checkpoint_dir="${TRAIN_LOGDIR}" \
  --eval_logdir="${EVAL_LOGDIR}" \
  --dataset_dir="${LANE_DATASET}" \
  --max_number_of_evaluations=1

python "${WORK_DIR}"/vis_lane.py \
  --logtostderr \
  --vis_split="val" \
  --dataset="lane_seg" \
  --model_variant="xception_65" \
  --atrous_rates=6 \
  --atrous_rates=12 \
  --atrous_rates=18 \
  --output_stride=16 \
  --decoder_output_stride=4 \
  --vis_crop_size=513 \
  --vis_crop_size=513 \
  --checkpoint_dir="${TRAIN_LOGDIR}" \
  --vis_logdir="${VIS_LOGDIR}" \
  --dataset_dir="${LANE_DATASET}" \
  --max_number_of_iterations=1

python "${WORK_DIR}"/export_model.py \
  --logtostderr \
  --checkpoint_path="${CKPT_PATH}" \
  --export_path="${EXPORT_PATH}" \
  --model_variant="xception_65" \
  --atrous_rates=6 \
  --atrous_rates=12 \
  --atrous_rates=18 \
  --output_stride=16 \
  --decoder_output_stride=4 \
  --num_classes=2 \
  --crop_size=513 \
  --crop_size=513 \
  --inference_scales=1.0
```
I alter the train_utils.py's exclude_list=['global_step',logits]
when i train the dataset,i error list:

Input checkpoint '/home/models/research/deeplab/datasets/lane_seg/exp/train_on_trainval_set/train/model.ckpt-10000' doesn't exist!

my questions is:

    when I vis the picture,The result is like:
    i think it only use pretrain of voc to inference,so the result is so bad,Is my opinion right?
![](http://img2.ph.126.net/5aK_IRuxMATmbE52dhRRkQ==/2990108677697661746.png)
![](http://img1.ph.126.net/uHDFeEL3U9BzbjvPfuqkmw==/6597626119542484963.png)
    my dataset mybe the data samples may be strongly biased to background,How can i finetune the weight of loss ?
    Your num_classes should be greater than the max pixel value in the images
    Someone said that should finetune loss's weight,what I shold do,scale label 1's weight to 10x,100x,or change to 0.1,0.2or change label 0's weight?what others i need to finetune?

Thanks for your help!

**Comment by austinmw** [comment]
Not sure about your other questions, but did you verify that you have a set of three 10,000 iteration ckpt files in the specified location?

**Comment by bleedingfight** [comment]
@austinmw thanks for your help,the checkpoint proble i find my checkpoint file isn't exist,I shold use deeplab initial model instead it?or just train it from scrach?

**Comment by bleedingfight** [comment]
@austinmw I try to train the models,but the loss oscillation about 0.8 even though I have tried it for 10w step,Have some suggestion to deal with it?(I have tried to decrease lr to 0.000001 but the loss as I just said)

**Comment by Citygity** [comment]
@bleedingfight  Hey,I'm trying to train my own dataset just like your experiment(2 CLASS include backgroud).But I can just  get mean_iou=1 after running eval.py and get a black output after  running vis.py.I think I've made some error during making my training data .So, would you mind sharing your experience on making biclass training data ?Thanks a lot !

**Comment by bleedingfight** [comment]
@Citygity 
```
 not_ignore_mask = tf.to_float(tf.equal(scaled_labels, 0)) * 1 + tf.to_float(tf.    equal(scaled_labels, 1)) * 500 + tf.to_float(tf.equal(scaled_labels, ignore_label))     * 0

```

**Comment by Citygity** [comment]
@bleedingfight thanks a lot ,I have added this part in train_utils.py  .However it doesn't work ,I get same result( mean_iou=1 and black output) .My dataset is spacenet(a high-resolution remote sensing imagery dataset) and I made the label image by [https://github.com/SpaceNetChallenge/utilities](url) .The label image was a PNG format image with 2 color(0 for backround and 122 for building).So I'd like to know how did you build your own data's label?thanks a lot !
Happy Dragon Boat Festival!
端午节快乐！

**Comment by Citygity** [comment]
@bleedingfight Your method works, thanks a lot, I got bad result too as follow.
![000192_image](https://user-images.githubusercontent.com/8197720/41540974-ec50adba-7343-11e8-855e-fa6c79b464fc.png)
![000192_prediction](https://user-images.githubusercontent.com/8197720/41540975-eca8496c-7343-11e8-85df-1b413f900a70.png)

**Comment by bleedingfight** [comment]
@Citygity maybe you. shold train it more step to see the reault. my result is bad that miou is 0.46,i. don't know why!

**Comment by Citygity** [comment]
@bleedingfight I‘m trying to train 40000 steps ,I think I'd be able to get the result tomorrow morning. I'd share with you if I get any trick training this model.

**Comment by bleedingfight** [comment]
@Citugity thanks for your share

**Comment by Citygity** [comment]
@bleedingfight I got miou_1.0[0.81052655] after 40000 steps training.I changed the parameter in ` not_ignore_mask = tf.to_float(tf.equal(scaled_labels, 0)) * 1 + tf.to_float(tf.    equal(scaled_labels, 1)) * 500 + tf.to_float(tf.equal(scaled_labels, ignore_label))     * 0` into ` not_ignore_mask = tf.to_float(tf.equal(scaled_labels, 0)) * 1 + tf.to_float(tf.    equal(scaled_labels, 1)) * 5 + tf.to_float(tf.equal(scaled_labels, ignore_label))     * 0`(changed 500 to 5)
the result is just as follow:
![000192_image](https://user-images.githubusercontent.com/8197720/41570430-62a97818-73a1-11e8-81da-885f7bd5a7d3.png)
![000192_prediction](https://user-images.githubusercontent.com/8197720/41570433-62fcbf82-73a1-11e8-8303-034bd5fde85a.png)

**Comment by weehe91325** [comment]
Hi

Currently I’m struggling with improving the results using deeplab trained on my own dataset.
I’ve trained deeplab successfully a few times using different pretrain models from the model zoo, all based on xception_65, but my results keep staying in the same miou range, somewhere around this interval [10, 11].
I have only one GPU at my disposal with 11GB GPU memory.
My dataset has 8 classes with various object sizes, from little to big, and is quite unbalanced.
Here are the ratios: [1, 4, 4, 17, 42, 36, 19, 20].
In my dataset I have 757 instances for training and 100 validation.
I’ve tried to adjust parameters like learning rate, last_layer_gradient_multiplier, weight decay.
I’ve also tried some kind of weighting using the above ratios in this formula

weights = tf.to_float(tf.equal(scaled_labels, 0)) * 1 + 
tf.to_float(tf.equal(scaled_labels, 1)) * 4 + 
tf.to_float(tf.equal(scaled_labels, 2)) * 4 + 
tf.to_float(tf.equal(scaled_labels, 3)) * 17 + 
tf.to_float(tf.equal(scaled_labels, 4)) * 42 + 
tf.to_float(tf.equal(scaled_labels, 5)) * 36 + 
tf.to_float(tf.equal(scaled_labels, 6)) * 19 + 
tf.to_float(tf.equal(scaled_labels, 7)) * 20 + 
tf.to_float(tf.equal(scaled_labels, ignore_label)) * 0.0

but it turns out the algorithm won’t converge.
I’ve trained without fine tuning the batch normalization parameters. Although I tried training those parameters with a 321 crop size in order to be able to fit a batch size of 12 in my GPU.
I’ve tried training on various sizes 321, 513, 769.
The point being I need some tips to figure out what I can do to improve those results.
What do you guys think? Do I need more data in order to increase my miou or hardware?

**Comment by austinmw** [comment]
Someone else can chime in and probably help more, but I don't know if batch normalization is worth the time it takes for very small batch sizes used in mask rcnn models? if you have the time you could plot a learning curve by repeating training with multiple data amounts to get a better idea if more data will help.

**Comment by trobr** [comment]
@Citygity Hi, I had the same problem when I was training the model, and my trian image is png format too,Did you transform the train image from png to jpeg format?

**Comment by kritiyer** [comment]
For others who have miou=1 and blank segmentation maps for binary classification, set the ignore_label parameter in segmentation_dataset.py to a number that is not present in your label images. My labels were 0 and 1 so I set `ignore_label=255` and my segmentation maps are no longer blank.

@trobr I used PNG images so I don't think you need to change them to jpeg

**Comment by trobr** [comment]
@kritiyer Thank you for your help!There ara 12 classification with background in my dataset,and i set the parameter _PASCAL_VOC_SEG_INFORMATION as shown below,but I get the problem too.And i have another confusion,Is the label image format in your dataset is np.uint8 or np.uint16.In my dataset,they are set to np.uint16,could this be the cause of the problem?I'm going to convert the label image format and retrain again.Thanks again for your suggest.
![image](https://user-images.githubusercontent.com/40254133/43180887-04bacf1c-900d-11e8-94d0-c2295e7a2462.png)

**Comment by kritiyer** [comment]
@trobr I used uint8 labels

**Comment by trobr** [comment]
@kritiyer year!I think I had find out the cause of this problem,when I convert label to uint8 and I have got expected results.Thanks very much for your help！

**Comment by Ertyoii** [comment]
Hi
@Citygity is there any way to calculate iou instead of mean_iou for this kind of binary classification task using this deeplabv3+? There seems to be no tensorflow API to do that.

**Comment by ZuxianShi** [comment]
@Citygity My dataset also is spacenet, the size is 2001*2001, but i got a very bad result. Could you give me some advice? Thank you very much!

**Comment by anandhupvr** [comment]
@bleedingfight  @Citygity Hey,I'm trying to train my own dataset just like your experiment(2 CLASS include backgroud) but i get blank output The label image was a PNG format image with 2 color(0 for backround and 1 for foreground) 

SEG_INFORMATION = DatasetDescriptor(
    splits_to_sizes={
        'train': 250, # number of file in the train folder
        'trainval': 255,
        'val': 5,
    },
    num_classes=2, # number of classes in your dataset
    ignore_label=255, # white edges that will be ignored to be class


` not_ignore_mask = tf.to_float(tf.equal(scaled_labels, 0)) * 1 + tf.to_float(tf. equal(scaled_labels, 1)) * 500 + tf.to_float(tf.equal(scaled_labels, ignore_label)) * 0`

what i did wrong stuck for 2, 3 days 

Thanks for your help!

and i run only train.py and  export_model.py and tested with frozeninference.pb 

![apricot_2](https://user-images.githubusercontent.com/23016323/52344967-fb472600-2a13-11e9-8841-0d0c5c7dde72.png)

**Comment by tensorflowbutler** [comment]
Hi There,
 We are checking to see if you still need help on this, as this seems to be considerably old issue. Please update this issue with the latest information, code snippet to reproduce your issue and error you are seeing.
 If we don't hear from you in the next 7 days, this issue will be closed automatically. If you don't need help on this issue any more, please consider closing this.
